data/blog/improving-nodejs-loader-performance.md

```diff
@@ -0,0 +1,180 @@
+
+## How does Node.js parses `package.json`?
+
+By default, `readPackage` calls an internal version `fs.readFileSync` to read the `package.json` file. This synchronous call returns a string from Node.js C++ layer, which later gets parsed using V8's `JSON.parse()` method. Depending on the validity of this JSON, Node.js checks and creates an object that's required for the remaining of the loaders to perform. These fields are `pkg.name`, `pkg.main`, `pkg.exports`, `pkg.imports` and `pkg.type`. If the JSON has faulty syntax, Node.js will throw an error and exit the process.
+
+The output of this function is later cached at an internal `Map` to avoid calling `readPackageScope` again for the same path. This cache is stored for the rest of the process lifetime.
+
+## Usage of `package.json` fields and the reader
+
+Before we dive into what optimizations we can do, let's see how Node.js uses these fields. The common use cases in Node.js codebase for parsing and re-using `package.json` fields are:
+
+- `pkg.exports` and `pkg.imports` are used to resolve different modules according to your input.
+- `pkg.main` is used to resolve the entry point of the application.
+- `pkg.type` is used to resolve the module format of the file.
+- `pkg.name` is used if there is a self referencing require/import.
+
+Additionally, Node.js supports an experimental version of `Subresource Integrity` checkwhich uses the result of this package.json to validate the integrity of the file.
+
+The most important usage is that, for every `require/import` call, Node.js needs to know the module format of the file. For example, if the user require's a NPM module that uses ESM on a CommonJS (CJS) application, Node.js will need to parse the `package.json` file of that module and throw an error if the NPM package is ESM.
+
+Because of all of these calls and usages across ESM and CJS loaders, `package.json` reader is one of the most important parts of Node.js loader implementation.
+
+## Optimizations
+
+### Optimizing caching layer
+
+In order to optimize the `package.json` reader performance, I first moved the caching layer to the C++ side to make the implementation be closer to the filesystem call as much as possible. This decision forced to parse the JSON file in C++. At this point, I had 2 options:
+
+- Use V8's `v8::JSON::Parse()` method which takes a `v8::String` as an input and returns a `v8::Value` as an output.
+- Use `simdjson` library to parse the JSON file.
+
+Since, the filesystem returns a string, converting that string into a `v8::String` just to retrieve the keys and values as a `std::string` didn't make sense. Therefore, I added `simdjson` as a dependency to Node.js and used it to parse the JSON file. This change enabled us to parse the JSON file in C++ and extract and return only the necessary fields to the JavaScript side, reducing the size of the input that needs to be serialized/deserialized.
```

Comment:
```suggestion
Since the filesystem returns a string, converting that string into a `v8::String` just to retrieve the keys and values as a `std::string` didn't make sense. Therefore, I added `simdjson` as a dependency to Node.js and used it to parse the JSON file. This change enabled us to parse the JSON file in C++ and extract and return only the necessary fields to the JavaScript side, reducing the size of the input that needs to be serialized/deserialized.
```